// pico-s3/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PicoS3 = void 0;
const client_1 = require("./client");
__exportStar(require("./client"), exports);
/**
 * A simple class wrapping of the pico s3 functions.
 */
class PicoS3 {
    /**
     * @param {S3RequestOptions} options The authentication details of the S3 bucket.
     *
     * You can simple set this to `true` to get the library to infer the authentication deails of the bucket from the environment variables.
     *
     * Here are the environment variables that need to be set:
     * ```
     * provider         = process.env.PICO_S3_CLOUD_PROVIDER
     * region           = process.env.PICO_S3_REGION
     * bucket           = process.env.PICO_S3_BUCKET
     * accessKeyId      = process.env.PICO_S3_ACCESS_KEY_ID
     * secretAccessKey  = process.env.PICO_S3_SECRET_ACCESS_KEY
     * ```
     */
    constructor(options) {
        if (typeof options === "boolean" && options === true) {
            this.options = {
                provider: process.env.PICO_S3_CLOUD_PROVIDER,
                region: process.env.PICO_S3_REGION,
                bucket: process.env.PICO_S3_BUCKET,
                accessKeyId: process.env.PICO_S3_ACCESS_KEY_ID,
                secretAccessKey: process.env.PICO_S3_SECRET_ACCESS_KEY,
            };
        }
        else
            this.options = options;
    }
    upload(options) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield client_1.upload(Object.assign(Object.assign({}, this.options), options));
        });
    }
    getObject(options) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield client_1.getObject(Object.assign(Object.assign({}, this.options), options));
        });
    }
    getObjectBuffer(options) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield client_1.getObjectBuffer(Object.assign(Object.assign({}, this.options), options));
        });
    }
    getObjectDataUrl(options) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield client_1.getObjectDataUrl(Object.assign(Object.assign({}, this.options), options));
        });
    }
    getObjectBinary(options) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield client_1.getObjectBinary(Object.assign(Object.assign({}, this.options), options));
        });
    }
    deleteObject(options) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield client_1.deleteObject(Object.assign(Object.assign({}, this.options), options));
        });
    }
    objectExists(options) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield client_1.objectExists(Object.assign(Object.assign({}, this.options), options));
        });
    }
    getObjectMetadata(options) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield client_1.getObjectMetadata(Object.assign(Object.assign({}, this.options), options));
        });
    }
    getObjectEtag(options) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield client_1.getObjectEtag(Object.assign(Object.assign({}, this.options), options));
        });
    }
    getTextFile(options) {
        return __awaiter(this, void 0, void 0, function* () {
            return yield client_1.getTextFile(Object.assign(Object.assign({}, this.options), options));
        });
    }
}
exports.PicoS3 = PicoS3;
