// core-types-graphql/dist/lib/gql-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isRequired = exports.gqlStripRequired = exports.gqlMaybeRequiredNode = exports.gqlListTypeNode = exports.gqlNamedTypeNode = exports.gqlUnionType = exports.gqlNameNode = void 0;
function gqlNameNode(name) {
    return { kind: 'Name', value: name };
}
exports.gqlNameNode = gqlNameNode;
function gqlUnionType({ name, description, types }) {
    return {
        kind: 'UnionTypeDefinition',
        name,
        ...(description ? { description } : {}),
        types,
    };
}
exports.gqlUnionType = gqlUnionType;
function gqlNamedTypeNode(name) {
    return { kind: 'NamedType', name: gqlNameNode(name) };
}
exports.gqlNamedTypeNode = gqlNamedTypeNode;
function gqlListTypeNode(type) {
    return { kind: 'ListType', type };
}
exports.gqlListTypeNode = gqlListTypeNode;
function gqlMaybeRequiredNode(type, required) {
    return !required ? type : { kind: 'NonNullType', type };
}
exports.gqlMaybeRequiredNode = gqlMaybeRequiredNode;
function gqlStripRequired(type) {
    return isRequired(type) ? type.type : type;
}
exports.gqlStripRequired = gqlStripRequired;
function isRequired(type) {
    return type.kind === 'NonNullType';
}
exports.isRequired = isRequired;
